cloud: early-boot storage providers

stevendanna opened the pull request:
This adds the concept of "early boot" external storage providers. Such providers can be constructed and used without depending on a working SQL or storage layer.

We want such a distinction to support using external storage providers from pebble itself. When starting a node with external SSTs, we would like to be able to open the providers for those SSTs without regard to where the server startup is.

Previously, we attempted to simply block until the required dependencies were available. However, this quickly revealed a deadlock at startup.  While we could potentially fix the deadlock, we do not want to have to be constantly vigilant for minor changes to pebble that could re-introduce such a deadlock.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114629)
<!-- Reviewable:end -->

msbutler commented on pkg/cloud/impl_registry.go:
```diff
@@ -380,3 +472,91 @@ type ReadWriterInterceptor interface {
 }
+
+// makeExternalStorage creates an ExternalStorage from the given config.
+func makeExternalStorage[T any](
```
nit: narrow the generic type? `s/T any/ T EarlyBootExternalStorageContext | ExternalStorageContext/r`

stevendanna replied:
Doneburger.

msbutler commented on pkg/cloud/impl_registry.go:
```diff
@@ -245,18 +337,18 @@ func makeLimiter(
 	s.rate.SetOnChange(sv, fn)
 	s.burst.SetOnChange(sv, fn)
-	fn(ctx)
+	fn(context.Background())
```
why not remove the context arg all together from `fn`?

dt replied:
`fn` is passed to `SetOnChange` and needs to adhere to its arg type.

msbutler commented on pkg/cloud/impl_registry.go:
```diff
@@ -380,3 +472,91 @@ type ReadWriterInterceptor interface {
+	return nil, errors.Errorf("unsupported external destination type: %s", dest.Provider.String())
+}
+
+func HijackProviderForTesting(
```
nit: s/Hijack/Replace/r :D

Perhaps also name the return func `cleanup` or provide a sentence in the docstring on what this function does

dt replied:
or `Intercept`

stevendanna replied:
Went with Replace.

dt replied:
Sorry, I know you _just_ replaced all these names, but I believe our convention is to prefix with `Testing` e.g. `TestingReplaceProvider`

msbutler commented on pkg/cloud/impl_registry.go:
```diff
@@ -95,34 +97,85 @@ func registerLimiterSettings(providerType cloudpb.ExternalStorageProvider) {
+		} else if provider.EarlyBootParseFn != nil {
+			confParsers[scheme] = func(_ ExternalStorageURIContext, u *url.URL) (cloudpb.ExternalStorage, error) {
+				return provider.EarlyBootParseFn(u)
+			}
```
also wonder if it's worth panicking if both `ParseFn` and `EarlyBootParseFn` are not nil

stevendanna replied:
Doneburger.

msbutler reviewed (commented):
This looks great! I left mostly nits. Also, "EarlyBoot" seems fine to me. We can also go with "NoSQL" or "LimitedDeps", but I don't feel strongly.

One q: this work is motivated to prevent deadlock on server startup. Is there a way we can unit test that this deadlock does not occur on base and tenant server startup?

msbutler commented:
It may also be worth spinning up a manual run of the Cloud Unit test suite on TC before merging this.

dt commented:
Could I tempt you into splitting this into two/theee commits: first adding the testing intercept and call the calls to it (as one or maybe as two commits), then the early boot providers change / changes to server startup?

stevendanna commented on pkg/ccl/backupccl/restore_online_test.go:
```diff
@@ -26,6 +27,8 @@ func TestOnlineRestoreBasic(t *testing.T) {
 	_, sqlDB, dir, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts, InitManualReplication)
 	defer cleanupFn()
+
+	defer nodelocal.HijackNodeLocalForTesting(t.TempDir())()
```
The normal nodelocal storage provider isn't available at early boot because it requires that you have a node ID and can send RPCs to other nodes.  Oneline restore requires a storage provider that is available before those things are true since we may need to reboot and read from the store very early in startup.

We could have made a whole different storage provider just for testing. And that may be been a bit cleaner, but this allows us to replace the nodelocal storage provider with an implementation that just serves files out of a temp directory without adding a new provider.

stevendanna commented on pkg/cloud/external_storage.go:
```diff
@@ -184,42 +200,50 @@ type ExternalStorageConstructor func(
-// NewExternalStorageAccessor creates an uninitialized ExternalStorageAccessor.
-func NewExternalStorageAccessor() *ExternalStorageAccessor {
-	return &ExternalStorageAccessor{ready: make(chan struct{})}
+type EarlyBootExternalStorageConstructor func(
```
Hrm. I'm all for better names as I dislike many of the names in this package including this one.

But, the reason why we have the two different names here is that typically "Factory" refers to a method that knows how to select between multiple implementations of a given interface for some configuration. 

For example, if you have an Interface A with implementations _f_ and _g_ an AFactory would be what is responsible for choosing between _f_ and _g_.  

Here the FromURIFactory is a factory because it knows how to take a URI and then route that to a specific ExternalStorage implementation.  The `Constructor` types is implemented by an ExternalStorage implementation and is what is _called by a factory_.

msbutler replied:
ah, thank you for the explanation. This makes sense. I had not realized Factory had a specific meaning.

stevendanna committed:
cloud: add nodelocal.ReplaceNodeLocalForTesting

This adds nodelocal.ReplaceNodeLocalForTesting which allows tests to
replace the production implementation of the nodelocal external
storage provider with one that reads and writes to given temporary
directory.

Epic: none
Release note: None

stevendanna committed:
backupccl: override nodelocal provider in restore tests

This commit is in preparation for a change that will make the
production implementation of the node local provider incompatible with
online restore.

Epic: none
Release note: None

stevendanna committed:
cloud: early-boot storage providers

This adds the concept of "early boot" external storage providers. Such
providers can be constructed and used without depending on a working
SQL or storage layer.

We want such a distinction to support using external storage providers
from pebble itself. When starting a node with external SSTs, we would
like to be able to open the providers for those SSTs without regard to
where the server startup is.

A subsequent commit will ensure that online restore only uses early
boot providers.

Epic: none
Release note: None

stevendanna committed:
server,backupccl: only allow early boot providers in online restore

When starting a node with external SSTs, we would like to be able to
open the providers for those SSTs without regard to where the server
startup is.

Previously, we attempted to simply block until the required
dependencies were available. However, this quickly revealed a
deadlock at startup.  While we could potentially fix the deadlock, we
do not want to have to be constantly vigilant for minor changes to
pebble that could re-introduce such a deadlock.

Here, we solve this by only allowing online restore and pebble to use
"early boot" external storage providers. such providers are available
immediately.

Epic: none
Release note: None

stevendanna commented:
> Could I tempt you into splitting this into two/theee commits: first adding the testing intercept and call the calls to it (as one or maybe as two commits), then the early boot providers change / changes to server startup?

Sure thing. I also split out the introduction of the new type or provider from its integration with server startup.

stevendanna commented:
I've looked into the remaining test failures and they appear to be unrelated flakes. I'm going.to merge this. If we see an increase in test flakes, we can revert.

bors r=dt,msbutler
